File: src/main/java/nl/dannyjelsma/soundboardbot/commands/music/PlaylistCommand.java
package nl.dannyjelsma.soundboardbot.commands.music;

import com.jagrosh.jdautilities.command.Command;
import com.jagrosh.jdautilities.command.CommandEvent;
import com.sedmelluq.discord.lavaplayer.player.AudioLoadResultHandler;
import com.sedmelluq.discord.lavaplayer.tools.FriendlyException;
import com.sedmelluq.discord.lavaplayer.track.AudioPlaylist;
import com.sedmelluq.discord.lavaplayer.track.AudioTrack;
import nl.dannyjelsma.soundboardbot.audio.SoundManager;
import nl.dannyjelsma.soundboardbot.music.Playlist;
import nl.dannyjelsma.soundboardbot.music.PlaylistManager;
import nl.dannyjelsma.soundboardbot.music.Song;
import nl.dannyjelsma.soundboardbot.result.PlaylistCreationResult;
import nl.dannyjelsma.soundboardbot.result.PlaylistLoadResult;
import nl.dannyjelsma.soundboardbot.result.PlaylistResult;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

public class PlaylistCommand extends Command {

    private static final PlaylistManager playlistManager = new PlaylistManager();

    public PlaylistCommand() {
        name = "playlist";
        help = "See =playlist help";
    }

    @Override
    protected void execute(CommandEvent event) {

        String[] args = event.getArgs().split(" ");

        switch (args[0].toLowerCase()) {
            case "create" -> createPlaylist(event);
            case "add" -> addToPlaylist(event);
            case "list" -> listPlaylist(event);
            case "play" -> playPlaylist(event);
            case "remove" -> removeFromPlaylist(event);
            case "deletelist" -> deletePlaylist(event);
            default -> event.getChannel().sendMessage("""
                    Valid commands:
                    =playlist create <name>
                    =playlist add <name> <url>
                    =playlist list <name>
                    =playlist play <name>
                    =playlist remove <name> <url/title>
                    =playlist deletelist <name>""").queue();
        }
    }

    private void deletePlaylist(CommandEvent event) {
        String[] args = event.getArgs().split(" ");

        if (args.length != 2) {
            event.getChannel().sendMessage("Incorrect arguments! Usage: =playlist deletelist <name>").queue();
            return;
        }

        PlaylistResult result = playlistManager.deletePlaylist(args[1]);

        if (!result.isSuccess()) {
            event.getChannel().sendMessage("Could not delete the playlist: " + result.getMessage()).queue();
        } else {
            event.getChannel().sendMessage("Successfully deleted the playlist!").queue();
        }
    }

    private void removeFromPlaylist(CommandEvent event) {
        String[] args = event.getArgs().split(" ");

        if (args.length < 3) {
            event.getChannel().sendMessage("Incorrect arguments! Usage: =playlist remove <name> <url/title>").queue();
            return;
        }

        String playList = args[1];
        String arg = String.join(" ", Arrays.copyOfRange(args, 2, args.length));
        ;
        PlaylistResult result = playlistManager.removeFromPlaylist(playList, arg);

        if (!result.isSuccess()) {
            result = playlistManager.removeFromPlaylistByTitle(playList, arg);

            if (!result.isSuccess()) {
                event.getChannel().sendMessage("Could not remove the song from the playlist: " + result.getMessage()).queue();
            } else {
                event.getChannel().sendMessage("Successfully removed the song from the playlist!").queue();
            }
        } else {
            event.getChannel().sendMessage("Successfully removed the song from the playlist!").queue();
        }
    }

    private void playPlaylist(CommandEvent event) {
        String[] args = event.getArgs().split(" ");

        if (args.length != 2) {
            event.getChannel().sendMessage("Incorrect arguments! Usage: =playlist list <name>").queue();
            return;
        }

        String playListName = args[1];
        PlaylistLoadResult loadResult = playlistManager.loadPlaylist(playListName);

        if (!loadResult.isSuccess()) {
            event.getChannel().sendMessage("Could not load the playlist: " + loadResult.getMessage()).queue();
            return;
        }

        Playlist playlist = loadResult.getPlaylist();
        List<Song> songs = playlist.getSongs();
        SoundManager manager = SoundManager.getInstance();
        List<AudioTrack> tracks = new ArrayList<>();

        if (SoundManager.getInstance().shouldShuffle()) {
            Collections.shuffle(songs);
        }

        for (Song song : songs) {
            if (manager.isInCache(song.getUrl())) {
                tracks.add(manager.getFromCache(song.getUrl()));
            } else {
                manager.getAudioPlayerManager().loadItemOrdered(manager, song.getUrl(), new AudioLoadResultHandler() {
                    @Override
                    public void trackLoaded(AudioTrack track) {
                        tracks.add(track);
                        manager.addToCache(song.getUrl(), track);
                    }

                    @Override
                    public void playlistLoaded(AudioPlaylist playlist) {
                        tracks.add(playlist.getTracks().get(0));
                        manager.addToCache(song.getUrl(), playlist.getTracks().get(0));
                    }

                    @Override
                    public void noMatches() {
                    }

                    @Override
                    public void loadFailed(FriendlyException exception) {
                    }
                });
            }
        }

        manager.getScheduler().queue(tracks);
        event.getChannel().sendMessage("Added " + tracks.size() + " songs to the queue!").queue();
    }

    private void listPlaylist(CommandEvent event) {
        String[] args = event.getArgs().split(" ");

        if (args.length != 2) {
            event.getChannel().sendMessage("Incorrect arguments! Usage: =playlist list <name>").queue();
            return;
        }

        String playListName = args[1];
        PlaylistLoadResult loadResult = playlistManager.loadPlaylist(playListName);

        if (!loadResult.isSuccess()) {
            event.getChannel().sendMessage("Could not load the playlist: " + loadResult.getMessage()).queue();
            return;
        }

        Playlist playlist = loadResult.getPlaylist();
        List<Song> songs = playlist.getSongs();
        StringBuilder songListBuilder = new StringBuilder();
        boolean isFirstMsg = true;

        for (Song song : songs) {
            int currentLength = songListBuilder.length();
            String songLine = song.getTitle() + " - `" + song.getUrl() + "`\n";

            if (currentLength + songLine.length() >= 1500) {
                String songList = songListBuilder.toString();

                if (isFirstMsg) {
                    songList = "**" + playListName + " contains " + songs.size() + " songs:" + "**\n" + songList;
                    isFirstMsg = false;
                }

                event.getChannel().sendMessage(songList).queue();
                songListBuilder = new StringBuilder();
            } else {
                songListBuilder.append(songLine);
            }
        }

        if (songListBuilder.length() > 0) {
            String songList = songListBuilder.toString();
            event.getChannel().sendMessage(songList).queue();
        }
    }

    private void addToPlaylist(CommandEvent event) {
        String[] args = event.getArgs().split(" ");

        if (args.length < 3) {
            event.getChannel().sendMessage("Incorrect arguments! Usage: =playlist add <name> <url/search query>").queue();
            return;
        }

        String playList = args[1];
        String query = String.join(" ", Arrays.copyOfRange(args, 2, args.length));
        PlaylistResult result = playlistManager.addToPlaylist(playList, query);

        if (!result.isSuccess()) {
            event.getChannel().sendMessage("Could not add the song to the playlist: " + result.getMessage()).queue();
        } else {
            event.getChannel().sendMessage("Successfully added the song to the playlist!").queue();
        }
    }

    private void createPlaylist(CommandEvent event) {
        String[] args = event.getArgs().split(" ");

        if (args.length < 2) {
            event.getChannel().sendMessage("Incorrect arguments! Usage: =playlist create <name>").queue();
            return;
        } else if (args.length > 2) {
            event.getChannel().sendMessage("The playlist name cannot contain spaces! Use underscores instead.").queue();
            return;
        }

        String name = event.getArgs().split(" ")[1];
        PlaylistCreationResult result = playlistManager.createPlaylist(name);

        if (!result.isSuccess()) {
            event.getChannel().sendMessage("Could not create a new playlist with the name `" + name + "`: " + result.getMessage()).queue();
        } else {
            event.getChannel().sendMessage("Successfully created a new playlist named `" + name + "`!").queue();
        }
    }
}

File: src/main/java/nl/dannyjelsma/soundboardbot/commands/music/PlaylistCommand.kt
package nl.dannyjelsma.soundboardbot.commands.music

import com.jagrosh.jdautilities.command.Command
import com.jagrosh.jdautilities.command.CommandEvent
import com.sedmelluq.discord.lavaplayer.player.AudioLoadResultHandler
import com.sedmelluq.discord.lavaplayer.tools.FriendlyException
import com.sedmelluq.discord.lavaplayer.track.AudioPlaylist
import com.sedmelluq.discord.lavaplayer.track.AudioTrack
import nl.dannyjelsma.soundboardbot.audio.SoundManager
import nl.dannyjelsma.soundboardbot.music.PlaylistManager
import java.util.*


class PlaylistCommand : Command() {

    private val playlistManager = PlaylistManager()

    init {
        name = "playlist"
        help = "See =playlist help"
    }

    override fun execute(event: CommandEvent) {
        val args = event.args.split(" ").toTypedArray()
        when (args[0].lowercase(Locale.getDefault())) {
            "create" -> createPlaylist(event)
            "add" -> addToPlaylist(event)
            "list" -> listPlaylist(event)
            "play" -> playPlaylist(event)
            "remove" -> removeFromPlaylist(event)
            "deletelist" -> deletePlaylist(event)
            else -> event.channel.sendMessage("Valid commands:\n" +
                    "=playlist create <name>\n" +
                    "=playlist add <name> <url>\n" +
                    "=playlist list <name>\n" +
                    "=playlist play <name>\n" +
                    "=playlist remove <name> <url/title>\n" +
                    "=playlist deletelist <name>")
                .queue()
        }
    }

    private fun deletePlaylist(event: CommandEvent) {
        val args = event.args.split(" ").toTypedArray()
        if (args.size != 2) {
            event.channel.sendMessage("Incorrect arguments! Usage: =playlist deletelist <name>").queue()
            return
        }

        val result = playlistManager.deletePlaylist(args[1])
        if (!result.isSuccess) {
            event.channel.sendMessage("Could not delete the playlist: " + result.message).queue()
        } else {
            event.channel.sendMessage("Successfully deleted the playlist!").queue()
        }
    }

    private fun removeFromPlaylist(event: CommandEvent) {
        val args = event.args.split(" ").toTypedArray()
        if (args.size < 3) {
            event.channel.sendMessage("Incorrect arguments! Usage: =playlist remove <name> <url/title>").queue()
            return
        }

        val playList = args[1]
        val arg = args.copyOfRange(2, args.size).joinToString(" ")
        var result = playlistManager.removeFromPlaylist(playList, arg)
        if (!result.isSuccess) {
            result = playlistManager.removeFromPlaylistByTitle(playList, arg)
            if (!result.isSuccess) {
                event.channel.sendMessage("Could not remove the song from the playlist: " + result.message).queue()
            } else {
                event.channel.sendMessage("Successfully removed the song from the playlist!").queue()
            }
        } else {
            event.channel.sendMessage("Successfully removed the song from the playlist!").queue()
        }
    }

    private fun playPlaylist(event: CommandEvent) {
        val args = event.args.split(" ").toTypedArray()
        if (args.size != 2) {
            event.channel.sendMessage("Incorrect arguments! Usage: =playlist list <name>").queue()
            return
        }

        val playListName = args[1]
        val loadResult = playlistManager.loadPlaylist(playListName)
        if (!loadResult.isSuccess) {
            event.channel.sendMessage("Could not load the playlist: " + loadResult.message).queue()
            return
        }

        val playlist = loadResult.playlist
        if (playlist != null) {
            val songs = playlist.songs

            val tracks: MutableList<AudioTrack> = ArrayList()
            if (SoundManager.shouldShuffle()) {
                songs.shuffle()
            }

            for (song in songs) {
                val audioTrack: AudioTrack? = SoundManager.getFromCache(song.url)

                if (audioTrack != null) {
                    tracks.add(audioTrack)
                } else {
                    SoundManager.audioPlayerManager.loadItemOrdered(SoundManager, song.url,
                        object : AudioLoadResultHandler {

                            override fun trackLoaded(track: AudioTrack) {
                                tracks.add(track)
                                SoundManager.addToCache(song.url, track)
                            }

                            override fun playlistLoaded(playlist: AudioPlaylist) {
                                tracks.add(playlist.tracks[0])
                                SoundManager.addToCache(song.url, playlist.tracks[0])
                            }

                            override fun noMatches() {}
                            override fun loadFailed(exception: FriendlyException) {}
                        })
                }
            }

            SoundManager.scheduler.queue(tracks)
            event.channel.sendMessage("Added " + tracks.size + " songs to the queue!").queue()
        }
    }

    private fun listPlaylist(event: CommandEvent) {
        val args = event.args.split(" ").toTypedArray()
        if (args.size != 2) {
            event.channel.sendMessage("Incorrect arguments! Usage: =playlist list <name>").queue()
            return
        }

        val playListName = args[1]
        val loadResult = playlistManager.loadPlaylist(playListName)
        if (!loadResult.isSuccess) {
            event.channel.sendMessage("Could not load the playlist: " + loadResult.message).queue()
            return
        }

        val playlist = loadResult.playlist

        if (playlist != null) {
            val songs = playlist.songs
            var songListBuilder = StringBuilder()
            var isFirstMsg = true
            for (song in songs) {
                val currentLength = songListBuilder.length
                val songLine = "${song.title} - `${song.url}`"
                if (currentLength + songLine.length >= 1500) {
                    var songList = songListBuilder.toString()
                    if (isFirstMsg) {
                        songList = "**$playListName contains ${songs.size} songs:**\n$songList"
                        isFirstMsg = false
                    }

                    event.channel.sendMessage(songList).queue()
                    songListBuilder = StringBuilder()
                } else {
                    songListBuilder.append(songLine)
                }
            }

            if (songListBuilder.isNotEmpty()) {
                val songList = songListBuilder.toString()
                event.channel.sendMessage(songList).queue()
            }
        }
    }

    private fun addToPlaylist(event: CommandEvent) {
        val args = event.args.split(" ").toTypedArray()
        if (args.size < 3) {
            event.channel.sendMessage("Incorrect arguments! Usage: =playlist add <name> <url/search query>").queue()
            return
        }

        val playList = args[1]
        val query = args.copyOfRange(2, args.size).joinToString(" ")
        val result = playlistManager.addToPlaylist(playList, query)
        if (!result.isSuccess) {
            event.channel.sendMessage("Could not add the song to the playlist: " + result.message).queue()
        } else {
            event.channel.sendMessage("Successfully added the song to the playlist!").queue()
        }
    }

    private fun createPlaylist(event: CommandEvent) {
        val args = event.args.split(" ").toTypedArray()
        if (args.size < 2) {
            event.channel.sendMessage("Incorrect arguments! Usage: =playlist create <name>").queue()
            return
        } else if (args.size > 2) {
            event.channel.sendMessage("The playlist name cannot contain spaces! Use underscores instead.").queue()
            return
        }

        val name = event.args.split(" ").toTypedArray()[1]
        val result = playlistManager.createPlaylist(name)
        if (!result.isSuccess) {
            event.channel.sendMessage("Could not create a new playlist with the name `" + name + "`: " + result.message)
                .queue()
        } else {
            event.channel.sendMessage("Successfully created a new playlist named `$name`!").queue()
        }
    }
}